compiler/plugins/target/ROCM/ROCMTarget.cpp

```diff
@@ -374,12 +361,30 @@ class ROCMTargetBackend final : public TargetBackend {
         llvmFunc->addFnAttr(
             "amdgpu-flat-work-group-size",
             (llvm::Twine("1, ") + llvm::Twine(flatWgSize)).str());
-        if (wavesPerEu > 0) {
-          llvmFunc->addFnAttr("amdgpu-waves-per-eu",
-                              std::to_string(wavesPerEu));
-        }
         if (targetArch.starts_with("gfx9"))
           addPreloadKernArgHint(llvmFunc);
+
+        // Set the amdgpu-waves-per-eu flag from config if given.
+        if (std::optional<IntegerAttr> attr =
+                getConfigIntegerAttr(targetAttr, "waves_per_eu")) {
+          llvmFunc->addFnAttr("amdgpu-waves-per-eu",
+                              std::to_string(attr->getValue().getSExtValue()));
+        }
+
+        // Override flags as given by target func attrs.
+        if (auto funcAttrs =
+                func->getAttrOfType<DictionaryAttr>("target_func_attrs")) {
+          for (NamedAttribute funcAttr : funcAttrs) {
+            auto value = dyn_cast<StringAttr>(funcAttr.getValue());
+            if (!value) {
```

Comment:
In the test I see this
```
target_func_attrs = {"amdgpu-waves-per-eu" = 2 : i64}
```
however here it looks like it's requiring the attribute to be a string. Does the below test actually pass here?

Reply:
Fixed. Test passes since the verification happens during serialization. This just moves the attribute.

Reply:
I see, makes sense. I think it would be nice to add some basic conversion here (i.e. if IntegerAttr, convert to string) because the strings look weird (shakes fist at llvm)